In [1]:
import numpy as np
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense

# Завантаження даних
max_features = 10000
maxlen = 200  # Максимальна довжина рецензії
batch_size = 32

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

# Заповнення або обрізання рецензій до максимальної довжини
x_train = pad_sequences(x_train, maxlen=maxlen)
x_test = pad_sequences(x_test, maxlen=maxlen)


17464789/17464789 [==============================] - 0s 0us/step


#Проста RNN

In [2]:
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, 32))
model_rnn.add(SimpleRNN(32))
model_rnn.add(Dense(1, activation='sigmoid'))

model_rnn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model_rnn.fit(x_train, y_train, epochs=5, batch_size=batch_size, validation_split=0.2)


Epoch 1/5
625/625 [==============================] - 35s 53ms/step - loss: 0.5759 - accuracy: 0.6776 - val_loss: 0.4668 - val_accuracy: 0.7882
Epoch 2/5
625/625 [==============================] - 32s 51ms/step - loss: 0.3914 - accuracy: 0.8332 - val_loss: 0.4431 - val_accuracy: 0.8084
Epoch 3/5
625/625 [==============================] - 31s 49ms/step - loss: 0.2532 - accuracy: 0.9017 - val_loss: 0.4309 - val_accuracy: 0.8288
Epoch 4/5
625/625 [==============================] - 31s 50ms/step - loss: 0.1393 - accuracy: 0.9524 - val_loss: 0.5192 - val_accuracy: 0.8144
Epoch 5/5
625/625 [==============================] - 30s 49ms/step - loss: 0.1079 - accuracy: 0.9653 - val_loss: 0.8209 - val_accuracy: 0.6614


#LSTM

In [3]:
model_lstm = Sequential()
model_lstm.add(Embedding(max_features, 32))
model_lstm.add(LSTM(32))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model_lstm.fit(x_train, y_train, epochs=5, batch_size=batch_size, validation_split=0.2)


Epoch 1/5
625/625 [==============================] - 55s 84ms/step - loss: 0.4267 - accuracy: 0.7977 - val_loss: 0.3074 - val_accuracy: 0.8704
Epoch 2/5
625/625 [==============================] - 47s 75ms/step - loss: 0.2341 - accuracy: 0.9112 - val_loss: 0.3101 - val_accuracy: 0.8742
Epoch 3/5
625/625 [==============================] - 49s 78ms/step - loss: 0.1765 - accuracy: 0.9368 - val_loss: 0.3311 - val_accuracy: 0.8684
Epoch 4/5
625/625 [==============================] - 47s 76ms/step - loss: 0.1419 - accuracy: 0.9488 - val_loss: 0.3559 - val_accuracy: 0.8610
Epoch 5/5
625/625 [==============================] - 53s 84ms/step - loss: 0.1083 - accuracy: 0.9599 - val_loss: 0.4318 - val_accuracy: 0.8662


#Bidirectional LSTM

In [5]:
model_bidirectional = Sequential()
model_bidirectional.add(Embedding(max_features, 32))
model_bidirectional.add(Bidirectional(LSTM(32)))
model_bidirectional.add(Dense(1, activation='sigmoid'))

model_bidirectional.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model_bidirectional.fit(x_train, y_train, epochs=5, batch_size=batch_size, validation_split=0.2)


Epoch 1/5
625/625 [==============================] - 95s 146ms/step - loss: 0.4125 - accuracy: 0.8049 - val_loss: 0.3383 - val_accuracy: 0.8560
Epoch 2/5
625/625 [==============================] - 84s 134ms/step - loss: 0.2384 - accuracy: 0.9090 - val_loss: 0.3305 - val_accuracy: 0.8548
Epoch 3/5
625/625 [==============================] - 87s 139ms/step - loss: 0.1787 - accuracy: 0.9340 - val_loss: 0.3315 - val_accuracy: 0.8710
Epoch 4/5
625/625 [==============================] - 85s 137ms/step - loss: 0.1359 - accuracy: 0.9516 - val_loss: 0.4259 - val_accuracy: 0.8614
Epoch 5/5
625/625 [==============================] - 82s 131ms/step - loss: 0.1077 - accuracy: 0.9632 - val_loss: 0.4188 - val_accuracy: 0.8626


#Глибока мережа з LSTM

In [4]:
model_deep_lstm = Sequential()
model_deep_lstm.add(Embedding(max_features, 32))
model_deep_lstm.add(LSTM(64, return_sequences=True))
model_deep_lstm.add(LSTM(64))
model_deep_lstm.add(Dense(1, activation='sigmoid'))

model_deep_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model_deep_lstm.fit(x_train, y_train, epochs=5, batch_size=batch_size, validation_split=0.2)


Epoch 1/5
625/625 [==============================] - 140s 218ms/step - loss: 0.4244 - accuracy: 0.7934 - val_loss: 0.3344 - val_accuracy: 0.8636
Epoch 2/5
625/625 [==============================] - 140s 224ms/step - loss: 0.2470 - accuracy: 0.9028 - val_loss: 0.3250 - val_accuracy: 0.8668
Epoch 3/5
625/625 [==============================] - 144s 230ms/step - loss: 0.1736 - accuracy: 0.9356 - val_loss: 0.3300 - val_accuracy: 0.8686
Epoch 4/5
625/625 [==============================] - 144s 230ms/step - loss: 0.1223 - accuracy: 0.9574 - val_loss: 0.3784 - val_accuracy: 0.8666
Epoch 5/5
625/625 [==============================] - 145s 232ms/step - loss: 0.0843 - accuracy: 0.9721 - val_loss: 0.4570 - val_accuracy: 0.8572


In [6]:
score_rnn = model_rnn.evaluate(x_test, y_test, verbose=0)
score_lstm = model_lstm.evaluate(x_test, y_test, verbose=0)
score_bidirectional = model_bidirectional.evaluate(x_test, y_test, verbose=0)
score_deep_lstm = model_deep_lstm.evaluate(x_test, y_test, verbose=0)

print("Проста RNN:")
print("Точність:", score_rnn[1])

print("\nLSTM:")
print("Точність:", score_lstm[1])

print("\nBidirectional LSTM:")
print("Точність:", score_bidirectional[1])

print("\nГлибока мережа з LSTM:")
print("Точність:", score_deep_lstm[1])


Проста RNN:
Точність: 0.6661199927330017

LSTM:
Точність: 0.849399983882904

Bidirectional LSTM:
Точність: 0.8498799800872803

Глибока мережа з LSTM:
Точність: 0.8472399711608887


Проста RNN показала найнижчу точність при класифікації рецензій на датасеті IMDb (приблизно 66.6%). Це свідчить про обмежену здатність простої RNN до моделювання складних залежностей в текстах, так як вона має проблему зі зниклими градієнтами.

LSTM та глибока мережа з LSTM показали значно кращі результати з точністю близько 85%. LSTM вже має можливість управляти довготривалими залежностями завдяки механізму довгих короткострокових пам'ятей (LSTM), тоді як глибока мережа з LSTM має ще більший потенціал для моделювання складних текстових зв'язків.

Bidirectional LSTM показала результат, подібний до звичайного LSTM, що вказує на те, що в даному випадку додавання бідірекціональності не значно покращило результати. Це може бути пов'язано з тим, що в деяких випадках зворотний контекст може вводити шум в модель.

Загалом, з точки зору точності, LSTM та глибока мережа з LSTM показали найкращі результати для класифікації рецензій у датасеті IMDb.